### DISCUSSION TITLE
Checking if self.groups is None in AsyncWebsocketConsumer causes ValueError

### DISCUSSION THREAD

**Discussion by utsdanielosborne** [django/channels#1647] [Q&A]
With reference to:
https://github.com/django/channels/blob/ece488b31b4e20a55e52948f21622da3e38223cb/channels/generic/websocket.py#L160
The check of self.groups during \__init__, to change it from None to an empty list 'if not set', causes a ValueError to be thrown by:
https://github.com/django/channels/blob/ece488b31b4e20a55e52948f21622da3e38223cb/channels/auth.py#L146
if self.groups is a property which tries to access scope['user'].

i.e.

    class Test (AsyncWebsocketConsumer):
    
        @property
        def groups (self):
            return [a_fn(self.scope['user']]

Because scope['user'] is only set during \__call__

I've had to work around this by catching ValueError in groups for when it's called from \__init__

It seems to me that the \__init__ check is only intended to ensure that self.groups returns a list, but does not need to access that list itself as it currently does.
Therefore this problem could be avoided perhaps by e.g defining a default groups property on AsyncWebsocketConsumer that can be overridden, and removing the two lines from \__init__.

i.e. 

    class AsyncWebsocketConsumer (AsyncConsumer):

        @property
        def groups (self):
            ## Override as required.
            return []
    
        ## Removed def __init__(self, *args, **kwargs):

**Comment by aryaniyaps** [comment]
Hello @utsdanielosborne 
Sorry for the late reply, but I think that the best way to set the groups attribute with reference to the scope's user would be to override the consumer's connect method. Here's an example.
```py
class Test (AsyncWebsocketConsumer):
    async def connect(self):
        # reject anonymous connections if the
        # scope's user shouldn't be anonymous
        await self.accept()
        self.groups = [a_fn(self.scope['user']]
```
This way, you're guaranteed that the request's user would be set.

**Comment by aryaniyaps** [comment]
As far as I know, the groups attribute is only for setting "predefined" group names, which don't reference any attributes of the scope itself.
